File: BackEnd/routes/achievements.js
/**
 * Achievements routes
 */

const express = require('express');
const router = express.Router();
const { authMiddleware } = require('../utils/auth');
const { getUserById, updateUser, getGameHistory } = require('../utils/db');

const ACHIEVEMENTS = [
  { id: 'first_win', name: 'Prima Vittoria', icon: '🎉', desc: 'Vinci la tua prima scommessa' },
  { id: 'high_roller', name: 'High Roller', icon: '💸', desc: 'Scommetti €100 in una mano' },
  { id: 'streak_3', name: 'Tre di Fila', icon: '🔥', desc: 'Vinci 3 scommesse consecutive' },
  { id: 'blackjack', name: 'Blackjack!', icon: '🃏', desc: 'Fai un blackjack naturale' },
  { id: 'jackpot', name: 'Jackpottaro', icon: '🎰', desc: 'Vinci più di €500 alle slot' },
  { id: 'poker_royal', name: 'Royal Flush', icon: '♠', desc: 'Ottieni un Royal Flush al poker' },
  { id: 'multi_bet', name: 'Combinatore', icon: '🎲', desc: 'Vinci una multi-scommessa (3+ eventi)' },
  { id: 'total_100', name: 'Centurione', icon: '💯', desc: 'Gioca 100 scommesse totali' },
  { id: 'balance_5000', name: 'Ricco Sfondato', icon: '🤑', desc: 'Raggiungi €5.000 di saldo' },
  { id: 'zero_hero', name: 'Zero Hero', icon: '🎰', desc: 'Indovina lo zero alla roulette' },
  { id: 'race_winner', name: 'Appassionato di Gare', icon: '🏁', desc: 'Vinci una corsa di cavalli' },
  { id: 'baccarat_tie', name: 'Puntatore Audace', icon: '💎', desc: 'Vinci puntando sul pareggio al baccarat' },
];

function checkAchievements(user) {
  const history = getGameHistory(user.id, 1000);
  const unlocked = [];

  // First win
  if (history.some(h => h.result === 'win')) {
    unlocked.push('first_win');
  }

  // High roller
  if (history.some(h => h.bet >= 100)) {
    unlocked.push('high_roller');
  }

  // Streak 3
  let streak = 0;
  for (const h of history) {
    if (h.result === 'win') {
      streak++;
      if (streak >= 3) {
        unlocked.push('streak_3');
        break;
      }
    } else {
      streak = 0;
    }
  }

  // Blackjack
  if (history.some(h => h.game === 'Blackjack' && h.result === 'win' && h.gain > h.bet)) {
    unlocked.push('blackjack');
  }

  // Jackpot
  if (history.some(h => h.game === 'Slot Machine' && h.gain > 500)) {
    unlocked.push('jackpot');
  }

  // Poker royal
  if (history.some(h => h.game === 'Video Poker' && h.gain >= h.bet * 800)) {
    unlocked.push('poker_royal');
  }

  // Multi bet
  if (history.some(h => h.game === 'Virtual Sports' && h.gain > h.bet * 5)) {
    unlocked.push('multi_bet');
  }

  // Total 100
  if (history.length >= 100) {
    unlocked.push('total_100');
  }

  // Balance 5000
  if (user.balance >= 5000) {
    unlocked.push('balance_5000');
  }

  // Zero hero
  if (history.some(h => h.game === 'Roulette' && h.gain > h.bet * 30)) {
    unlocked.push('zero_hero');
  }

  // Race winner
  if (history.some(h => h.game === 'Corse Cavalli' && h.result === 'win')) {
    unlocked.push('race_winner');
  }

  // Baccarat tie
  if (history.some(h => h.game === 'Baccarat' && h.gain >= h.bet * 8)) {
    unlocked.push('baccarat_tie');
  }

  return [...new Set(unlocked)];
}

// ── GET ACHIEVEMENTS ──
router.get('/', authMiddleware, (req, res) => {
  try {
    const user = getUserById(req.user.id);
    if (!user) return res.status(404).json({ error: 'User not found', status: 404 });

    const unlockedIds = checkAchievements(user);
    const achievements = ACHIEVEMENTS.map(a => ({
      ...a,
      unlocked: unlockedIds.includes(a.id)
    }));

    res.json({
      ok: true,
      achievements,
      unlockedCount: unlockedIds.length,
      totalCount: ACHIEVEMENTS.length
    });
  } catch (e) {
    console.error('[ACHIEVEMENTS] Error:', e);
    res.status(500).json({ error: 'Internal server error', status: 500 });
  }
});

module.exports = router;

File: BackEnd/src/stats/Achievements.java
package stats;

import core.*;
import java.util.*;

public class Achievements {
    
    public static class Achievement {
        public String id;
        public String name;
        public String icon;
        public String description;
        
        public Achievement(String id, String name, String icon, String description) {
            this.id = id;
            this.name = name;
            this.icon = icon;
            this.description = description;
        }
    }
    
    private static final List<Achievement> ALL_ACHIEVEMENTS = Arrays.asList(
        new Achievement("first_win", "Prima Vittoria", "🎉", "Vinci la tua prima scommessa"),
        new Achievement("high_roller", "High Roller", "💸", "Scommetti €100 in una mano"),
        new Achievement("streak_3", "Tre di Fila", "🔥", "Vinci 3 scommesse consecutive"),
        new Achievement("blackjack", "Blackjack!", "🃏", "Fai un blackjack naturale"),
        new Achievement("jackpot", "Jackpottaro", "🎰", "Vinci più di €500 alle slot"),
        new Achievement("poker_royal", "Royal Flush", "♠", "Ottieni un Royal Flush al poker"),
        new Achievement("multi_bet", "Combinatore", "🎲", "Vinci una multi-scommessa"),
        new Achievement("total_100", "Centurione", "💯", "Gioca 100 scommesse totali"),
        new Achievement("balance_5000", "Ricco Sfondato", "🤑", "Raggiungi €5.000 di saldo"),
        new Achievement("zero_hero", "Zero Hero", "🎰", "Indovina lo zero alla roulette"),
        new Achievement("race_winner", "Appassionato di Gare", "🏁", "Vinci una corsa di cavalli"),
        new Achievement("baccarat_tie", "Puntatore Audace", "💎", "Vinci puntando sul pareggio al baccarat")
    );
    
    public static void checkAndUnlock(User user) {
        List<GameRecord> history = Database.getGameHistory(user.getId());
        
        // First win
        if (history.stream().anyMatch(GameRecord::isWin)) {
            user.addAchievement("first_win");
        }
        
        // High roller
        if (history.stream().anyMatch(h -> h.getBet() >= 100)) {
            user.addAchievement("high_roller");
        }
        
        // Streak 3
        int streak = 0;
        for (GameRecord h : history) {
            if (h.isWin()) {
                streak++;
                if (streak >= 3) {
                    user.addAchievement("streak_3");
                    break;
                }
            } else {
                streak = 0;
            }
        }
        
        // Jackpot
        if (history.stream().anyMatch(h -> h.getGame().equals("Slot Machine") && h.getGain() > 500)) {
            user.addAchievement("jackpot");
        }
        
        // Total 100
        if (history.size() >= 100) {
            user.addAchievement("total_100");
        }
        
        // Balance 5000
        if (user.getSaldo() >= 5000) {
            user.addAchievement("balance_5000");
        }
        
        Database.saveUsers();
    }
    
    public static void displayAchievements(User user) {
        System.out.println("\n╔════════════════════════════════════════╗");
        System.out.println("║         🏅 TRAGUARDI SBLOCCATI         ║");
        System.out.println("╠════════════════════════════════════════╣");
        
        List<String> unlockedIds = user.getAchievements();
        int count = 0;
        
        for (Achievement ach : ALL_ACHIEVEMENTS) {
            if (unlockedIds.contains(ach.id)) {
                System.out.println("║ ✅ " + ach.icon + " " + ach.name);
                System.out.println("║    " + ach.description);
                count++;
            }
        }
        
        System.out.println("╠════════════════════════════════════════╣");
        System.out.println("║ Totale: " + count + "/" + ALL_ACHIEVEMENTS.size());
        System.out.println("╚════════════════════════════════════════╝\n");
    }
}
